### ISSUE TITLE
[gardening] x64 AOT compilation in debug mode takes very long for co19/LibTest/collection/ListBase/ListBase_class_A01_t02

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#54974]
TL;DR:

```
   - 88.10% dart::CallSiteInliner::InlineCalls()                                                                                                                                                                                                    ▒
      - 87.85% dart::CallSiteInliner::InlineStaticCalls()                                                                                                                                                                                           ▒
         - 87.67% dart::CallSiteInliner::TryInlining(dart::Function const&, dart::Array const&, dart::InlinedCallData*, bool)                                                                                                                       ▒
            - 87.65% dart::CallSiteInliner::TryInliningImpl(dart::Function const&, dart::Array const&, dart::InlinedCallData*, bool)                                                                                                                ▒
               - 82.04% dart::FlowGraphChecker::VisitBlocks()                                                                                                                                                                                       ▒
                  - 82.00% dart::FlowGraphChecker::VisitInstructions(dart::BlockEntryInstr*)                                                                                                                                                        ▒
                     - 70.13% dart::Instruction::GetBlock()    
```

AOT compilation for x64 in debug mode spends a long time in the inliner and flow graph checker in the specific test `co19/LibTest/collection/ListBase/ListBase_class_A01_t02`.

We should consider speeding up the flow graph checker.

=================================

The tests
```
co19/LibTest/collection/ListBase/ListBase_class_A01_t02 Timeout (expected Pass)
```
are failing on configurations
```
vm-aot-win-debug-x64
```

It's also flaking on:

* vm-aot-linux-debug-x64c
* vm-aot-win-debug-x64c

Current results: https://dart-current-results.web.app/#/filter=co19/LibTest/collection/ListBase/ListBase_class_A01_t02&showAll

```
--- Command "vm_compile_to_kernel" (took 03.000893s):
set DART_CONFIGURATION=DebugX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\DebugX64\crashpad_handler.exe & C:\b\s\w\ir\pkg\vm\tool\gen_kernel.bat --aot --platform=out/DebugX64/vm_platform_strong.dill -o C:\b\s\w\ir\out\DebugX64\generated_compilations\vm-aot-win-debug-x64\short12422_istBase_ListBase_class_A01_t02\out.dill C:\b\s\w\ir\tests\co19\src\LibTest\collection\ListBase\ListBase_class_A01_t02.dart -Dtest_runner.configuration=vm-aot-win-debug-x64 --packages=C:\b\s\w\ir\.dart_tool\package_config.json -Ddart.vm.product=false --sound-null-safety

exit code:
0

--- Command "precompiler" (took 16:00.000009s):
set DART_CONFIGURATION=DebugX64 & set DART_SUPPRESS_WER=1 & set DART_CRASHPAD_HANDLER=C:\b\s\w\ir\out\DebugX64\crashpad_handler.exe & out\DebugX64\gen_snapshot --snapshot-kind=app-aot-elf --elf=C:/b/s/w/ir/out/DebugX64/generated_compilations/vm-aot-win-debug-x64/short12422_istBase_ListBase_class_A01_t02/out.aotsnapshot --loading-unit-manifest=C:/b/s/w/ir/out/DebugX64/generated_compilations/vm-aot-win-debug-x64/short12422_istBase_ListBase_class_A01_t02/ignored.json --sound-null-safety -Dtest_runner.configuration=vm-aot-win-debug-x64 --ignore-unrecognized-flags --packages=C:\b\s\w\ir\.dart_tool\package_config.json C:\b\s\w\ir\out\DebugX64\generated_compilations\vm-aot-win-debug-x64\short12422_istBase_ListBase_class_A01_t02\out.dill

exit code:
1

diagnostics:
win_sdk_path not found

--- Re-run this test:
python3 tools/test.py -n vm-aot-win-debug-x64 co19/LibTest/collection/ListBase/ListBase_class_A01_t02
```

I'm not sure why the test is timing out, it finishes rather quickly, even in debug mode:

```
$ time xcodebuild/DebugARM64/dart tests/co19/src/LibTest/collection/ListBase/ListBase_class_A01_t02.dart

real	0m1.954s
user	0m1.607s
sys	0m0.114s
```

The flakiness is related to timeouts (and one single crash). [flakiness dashboard link](https://dashboards.corp.google.com/_0d7e1e34_2c58_4d29_bf12_cf7bed10396d?fb=name:in:co19%2FLibTest%2Fcollection%2FListBase%2FListBase_class_A01_t02&fh=eJytjrEKwzAMRP9FsyeTpfmDzhlDBuGorcCWjSsXSsm_x6KBls4dn0_vzvMLeIURBBOBg5BjS_JhfRbqdKdIQTuvqDgVCmdzvD8NgzfrhlXtaV4cpBaVp7cwam3UW1ij1cDmjrmQ5cLXVlE5y_fub3B8gKW0P-w7qJTyg_qpJduyA9MLUsM.)

Last linux debug x64 hang log:

```
--- Command "vm_compile_to_kernel" (took 04.000265s):
DART_CONFIGURATION=DebugX64 /b/s/w/ir/pkg/vm/tool/gen_kernel --aot --platform=out/DebugX64/vm_platform_strong.dill -o /b/s/w/ir/out/DebugX64/generated_compilations/vm-aot-linux-debug-x64/tests_co19_src_LibTest_collection_ListBase_ListBase_class_A01_t02/out.dill /b/s/w/ir/tests/co19/src/LibTest/collection/ListBase/ListBase_class_A01_t02.dart -Dtest_runner.configuration=vm-aot-linux-debug-x64 --packages=/b/s/w/ir/.dart_tool/package_config.json -Ddart.vm.product=false --sound-null-safety

exit code:
0

--- Command "precompiler" (took 16:00.000762s):
DART_CONFIGURATION=DebugX64 out/DebugX64/gen_snapshot --snapshot-kind=app-aot-assembly --assembly=/b/s/w/ir/out/DebugX64/generated_compilations/vm-aot-linux-debug-x64/tests_co19_src_LibTest_collection_ListBase_ListBase_class_A01_t02/out.S --sound-null-safety -Dtest_runner.configuration=vm-aot-linux-debug-x64 --ignore-unrecognized-flags --packages=/b/s/w/ir/.dart_tool/package_config.json /b/s/w/ir/out/DebugX64/generated_compilations/vm-aot-linux-debug-x64/tests_co19_src_LibTest_collection_ListBase_ListBase_class_A01_t02/out.dill

exit code:
1

diagnostics:
Process list including children: [120709]
Trying to capture stack trace for pid 120709
PID 120709 - process
TID 120709:
#0  0x000055e341f11240 dart::Instruction::AsBlockEntry() const
#1  0x000055e3422dc9a5 dart::Instruction::GetBlock()
#2  0x000055e3422ddc0a dart::Instruction::CanEliminate(dart::BlockEntryInstr const*) const
#3  0x000055e3423929b5 dart::DeadCodeElimination::EliminateDeadCode(dart::FlowGraph*)
#4  0x000055e3423bd980 dart::CompilerPass_DCE::DoBody(dart::CompilerPassState*) const
#5  0x000055e3423bc7f6 dart::CompilerPass::Run(dart::CompilerPassState*) const
#6  0x000055e3423bcee1 dart::CompilerPass::RunPipeline(dart::CompilerPass::PipelineMode, dart::CompilerPassState*)
#7  0x000055e34226979a dart::PrecompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)
#8  0x000055e34226a584 dart::PrecompileFunctionHelper(dart::Precompiler*, dart::CompilationPipeline*, dart::Function const&, bool)
#9  0x000055e342264c2c dart::Precompiler::CompileFunction(dart::Precompiler*, dart::Thread*, dart::Zone*, dart::Function const&)
#10 0x000055e34226360e dart::Precompiler::ProcessFunction(dart::Function const&)
#11 0x000055e34225dc24 dart::Precompiler::Iterate()
#12 0x000055e34225a4ec dart::Precompiler::DoCompileAll()
#13 0x000055e342258b16 dart::Precompiler::CompileAll()
#14 0x000055e342498e8e Dart_Precompile
#15 0x000055e341bcd852 dart::bin::main(int, char**)
#16 0x00007fc6be706083 __libc_start_main
#17 0x000055e341bca659 _start
TID 120710:
#0  0x00007fc6be80168e epoll_wait
#1  0x000055e341bd2145 dart::bin::EventHandlerImplementation::Poll(unsigned long)
#2  0x000055e341bfa4c2 dart::bin::ThreadStart(void*)
#3  0x00007fc6bea2b609 start_thread
#4  0x00007fc6be801353 __clone

--- Re-run this test:
python3 tools/test.py -n vm-aot-linux-debug-x64 co19/LibTest/collection/ListBase/ListBase_class_A01_t02
```

So it only hangs(?) on x64 and x64c. Only in AOT.

Last seen new results on flakiness dashboard: 2023-04-04 2023-04-05 for linux. So this has been flaky for a while.

I'll try to repro it locally.

Edit: AOT compilation indeed takes very long for this test!

**Comment by dcharkes** [comment]
Likely offending chain:

```
   - 88.10% dart::CallSiteInliner::InlineCalls()                                                                                                                                                                                                    ▒
      - 87.85% dart::CallSiteInliner::InlineStaticCalls()                                                                                                                                                                                           ▒
         - 87.67% dart::CallSiteInliner::TryInlining(dart::Function const&, dart::Array const&, dart::InlinedCallData*, bool)                                                                                                                       ▒
            - 87.65% dart::CallSiteInliner::TryInliningImpl(dart::Function const&, dart::Array const&, dart::InlinedCallData*, bool)                                                                                                                ▒
               - 82.04% dart::FlowGraphChecker::VisitBlocks()                                                                                                                                                                                       ▒
                  - 82.00% dart::FlowGraphChecker::VisitInstructions(dart::BlockEntryInstr*)                                                                                                                                                        ▒
                     - 70.13% dart::Instruction::GetBlock()                                                                                                                                                                                         ▒
                          dart::Instruction::AsBlockEntry() const       
```

**Comment by dcharkes** [comment]
FYI: @sstrickl @rmacnak-google @mkustermann 

Should we consider having a pointer to the surrounding block in every instruction?

Or maybe less aggressively running the flow graph checker after every try inlining and only running it after the inlining pass is done?

Any other suggestions.

**Comment by mkustermann** [comment]
I think this is a duplicate of https://github.com/dart-lang/sdk/issues/21708

**Comment by whesse** [comment]
Because this is causing deflaking times of up to an hour when deflaking just this single test, and the test is flakily timing-out, I am skipping this test on debug aot with SkipSlow in the co19-runtime.status file. The precompiler takes between 10 and 15 minutes to generate the snapshot.

Work on this is being tracked in issue #21708
